text/0502_aws-vpclattice.md

```diff
@@ -0,0 +1,1026 @@
+```typescript
+/**
+ * Create a vpc lattice service network.
+ * Implemented by `ServiceNetwork`.
+ */
+export interface IServiceNetwork extends core.IResource {
+
+  /**
+  * The Amazon Resource Name (ARN) of the service network.
+  */
+  readonly serviceNetworkArn: string;
+  /**
+   * The Id of the Service Network
+   */
+  readonly serviceNetworkId: string;
+  /**
+   * Is this an imported serviceNetwork
+   */
+  readonly imported: boolean;
+  /**
+   * Add Lattice Service
+   */
+  addService(props: AddServiceProps): void;
+  /**
+   * Associate a VPC with the Service Network
+   */
+  associateVPC(props: AssociateVPCProps): void;
+}
+```
+
+The Construct will consume ServiceNetworkProps.
+
+
+```typescript
+export interface ServiceNetworkProps {
+
+  /** The name of the Service Network. If not provided Cloudformation will provide
+   * a name
+   * @default cloudformation generated name
+   */
+  readonly name?: string;
+
+  /**
+   * The type of  authentication to use with the Service Network. The only method avaialble is
+   * AWS_IAM, however in anticipation, it is expected that other types will be added. 
+   * @default 'AWS_IAM'
+   */
+  readonly authType?: AuthType | undefined;
+
+  /**
+   * Logging destinations
+   * @default: no logging
+   */
+  readonly loggingDestinations?: LoggingDestination[];
+
+  /**
+   * Lattice Services that are assocaited with this Service Network
+   * @default no services are associated with the service network
+   */
+  readonly services?: IService[] | undefined;
+
+  /**
+   * Vpcs that are associated with this Service Network
+   * @default no vpcs are associated
+   */
+  readonly vpcs?: ec2.IVpc[] | undefined;
+  /**
+   * An access mode provides 'pre-canned' conditions for the policy
+   * @default no conditions
+   */
+  readonly accessmode?: ServiceNetworkAccessMode | undefined;
+}
+
+```
+`ServiceNetwork` will implement statics in addition to those in IServiceNetwork, to allow the import of the servicenetwork in other stacks ( potentially cross account ). `fromName` provides a lookup to obtain the Id, using a custom resource.  This provides a way to pass a concrete value between cross account stacks.  
+
+```typescript
+public static fromId(scope: constructs.Construct, id: string, props: ImportedServiceNetworkProps ): IServiceNetwork {
+    return new ImportedServiceNetwork(scope, id, props);
+}
+
+public static fromName(scope: constructs.Construct, id: string, serviceNetworkName: string ): IServiceNetwork {
+    return new ImportedServiceNetwork(scope, id, { serviceNetworkName: serviceNetworkName });
+  }
+```
+
+`ServiceNetwork` will implement methods for a variety of tasks;
+
+
+```typescript
+// adds an iam statement to the the authpolicy. Use this to create policy that is bespoke to this application 
+addStatementToAuthPolicy(statement: iam.PolicyStatement)
+```
+
+```typescript
+// applies the AuthPolicy that has been created, to the serviceNetwork. 
+applyAuthPolicyToServiceNetwork()
```

Comment:
It is possible ( and valid ) to create a servicenetwork ( or service ) that does not have a policy.. it essentially permits anything to use the service ( authenticated  or unauthed ).

the Policy gets applied to a service network or a service,  via a AWS::VpcLattice::AuthPolicy .. 

Its not directly on the ServiceNetwork, or Service. 

When we create rules for the listener, we have all the properties to allow create statements for the auth service policy..   We add add those statements... 
`this.service.authPolicy.addStatements(policyStatement)`

All would be fine, if the service / service network  referered to an IAM policy that had been created.  Unfourantly it does not.    The API ( and CF follows ), takes an IAM Like JSON string..     ( Why, this is madening! ) 

The reason why I have this applyAuthPolicy, is that until all the rules, and potentially any other manual statments are added, as well,   we cant' call it.    

This has caused me a lot of pain, and This felt like the best way to deal with it, but i'm open to other suggestions.

Reply:
Ok, this just took a bit more time to resolve.  

The rules can get passed as props to the ServiceNetwork ( and service ), rather than adding them afterwards using a method.  This was flawed and messy.  Now the auth policy will get associated when it needs to, without needing the user to add this to their code.